src/common/linear_assignment_optimizer.h

```diff
@@ -156,3 +158,129 @@ void auctionOptimization(M &value_matrix, std::array<int, N> &assignment, int si
 }
 
 }  // linear_assignment
+
+class LinearAssignmentCache {
+public:
+	static constexpr LinearAssignmentCache* kNotProvidedCache = nullptr;
+
+	class OptimizerInputOutput {
+	public:
+		OptimizerInputOutput()
+		    : target_slice_(Goal::Slice::Type(1)),
+		      source_slice_(Goal::Slice::Type(1)),
+		      assignemt_() {}
+
+		template <std::size_t N>
+		bool checkMatchAndGetResult(Goal::Slice &target_slice,
+		                            Goal::Slice &source_slice,
+		                            std::array<int, N> &assignemt) {
+			if (target_slice.getType() != target_slice_.getType() ||
+			    source_slice.getType() != source_slice_.getType()) {
+				return false;
+			}
+			assert(target_slice.size() == target_slice_.size());
+			assert(source_slice.size() == source_slice_.size());
+
+			for (int i = 0; i < target_slice.size(); i++) {
+				detail::Slice::PartProxy part = target_slice[i];
+				detail::Slice::PartProxy part_ = target_slice_[i];
+				if (part.size() != part_.size()) { return false; }
+				auto part_it_ = part_.begin();
```

Comment:
this iterator is for cached part so maybe we just call it cached_it ?
```suggestion
				auto cached_it = part_.begin();
```